### ISSUE TITLE
Preventing the generation of referenced types

### ISSUE THREAD

**Issue by andyws** [RSuter/NJsonSchema#414]
I have a set of Json Schema provided to me.  The set is complete and all references are resolved within the set.  Is there a way I can use the CS/TS generator to not generate types referenced in a schema?  Those types are themselves generated and I end up with multiple definitions of the same type.

**Comment by RSuter** [comment]
You can use the setting ExcludedTypeNames to exclude some types from being generated.

**Comment by andyws** [comment]
Hi Rico,

Thanks for responding.  Given a schema how can I determine the set of referenced types which I would then add to ExcludedTypeNames?

**Comment by RSuter** [comment]
You only need the root/entry schema to be generated?

**Comment by andyws** [comment]
Exactly, with one caveat.  Where a type has components which are defined 'locally' ( ie not defined in an external schema and which have no role outside of being a part of the root type )  I wish to generate both the root and its local components.

**Comment by gzak** [comment]
As a follow up, if there's a json file whose json-schema _only_ contains a `definitions` section without a root schema, it would be nice if it generated the types from the `definitions` section there for that file (in conjunction with _not_ generating it in each consumer of that file). Right now, if the root schema doesn't define a type, then nothing gets outputed, even if there is a `definitions` section.

Of course this is all contingent on whether or not a root object without a type definition is actually legal json-schema.

**Comment by gzak** [comment]
Basically, what me and I bet @andyws are getting at is: imaging you're building a repository of json-schemas from which you want to generate model classes automatically, and also you want to enable some amount of type sharing between types in this repo - i.e. I don't want to define the `Car` type again and again everywhere it gets used in the repository.

**Comment by RSuter** [comment]
1. Generating all schemas from definitions should be simple
2. We have to find a way how to configure the exclusions... Based on what? Only generate classes for schemas in the same file? Or based on the keys namespace?

**Comment by gzak** [comment]
Another possibility is to generate the root object as the main class, and then all helper classes as internal classes to the main class - that way you can generate the helper classes again and again, even though the definition is shared. Something like this:

    public class MainClass
    {
        public HelperClass SomeProperty { get; set; }

        public class HelperClass { ... }
    }

It's not quite as nice as what @andyws and I are proposing, since you then can't assign objects of the same "type" across classes, you'd have to copy them with something like [AutoMapper](http://automapper.org/). But it's something, at least.

**Comment by gzak** [comment]
2. Maybe just make it convention-based, and enable this convention via some parameter to the generator? I think the convention would be: do not generate externally resolved references, only generate local references. And for "empty" files containing only a `definitions` section, well these are technically local to that file so generate them there.

**Comment by RSuter** [comment]
Ok two settings

- GenerateDefinitionSchemas (default: false)
- IgnoreExternalSchemas

Im not sure if this information is preserved (is external) while deserializing a schema (FromJsonAsync)

**Comment by andyws** [comment]
@RSuter, @gzak Thanks for running with this.  As soon as the two settings are available I'll take time to work with it.
